spec/services/course_reporting_service_spec.rb

```diff
@@ -0,0 +1,166 @@
+  let(:closed_courses_program_type_scope) { class_double(Course) }
+  let(:closed_courses_study_mode_scope) { class_double(Course) }
+  let(:closed_courses_qualification_scope) { class_double(Course) }
+  let(:closed_courses_is_send_scope) { class_double(Course) }
+  let(:closed_course_subjects) { class_double(CourseSubject) }
+  let(:closed_course_subjects_grouped) { class_double(CourseSubject) }
+
+  let(:courses_scope) { class_double(Course) }
+  let(:findable_courses_scope) { class_double(Course) }
+
+  let(:open_courses_scope) { class_double(Course) }
+  let(:open_courses_provider_type_scope) { class_double(Course) }
+  let(:open_courses_program_type_scope) { class_double(Course) }
+  let(:open_courses_study_mode_scope) { class_double(Course) }
+  let(:open_courses_qualification_scope) { class_double(Course) }
+  let(:open_courses_is_send_scope) { class_double(Course) }
+  let(:open_course_subjects) { class_double(CourseSubject) }
+  let(:open_course_subjects_grouped) { class_double(CourseSubject) }
+
+  let(:distinct_courses_scope) { class_double(Course) }
+
+  let(:courses_count) { 100 }
+  let(:findable_courses_count) { 60 }
+  let(:open_courses_count) { 40 }
+  let(:closed_courses_count) { 20 }
+
+  let(:expected) do
+    {
+      total: {
+        all: courses_count,
+        non_findable: courses_count - findable_courses_count,
+        all_findable: findable_courses_count,
+      },
+      findable_total: {
+        open: open_courses_count,
+        closed: closed_courses_count,
+      },
+      provider_type: {
+        open: {
+          scitt: 1, lead_school: 2, university: 3, unknown: 4, invalid_value: 5
+        },
+        closed: {
+          scitt: 0, lead_school: 0, university: 0, unknown: 0, invalid_value: 0
+        },
+      },
+      program_type: {
+        open: {
+          higher_education_programme: 1, school_direct_training_programme: 2,
+          school_direct_salaried_training_programme: 3, scitt_programme: 4,
+          pg_teaching_apprenticeship: 5
+        },
+        closed: {
+          higher_education_programme: 0, school_direct_training_programme: 0,
+          school_direct_salaried_training_programme: 0, scitt_programme: 0,
+          pg_teaching_apprenticeship: 0
+        },
+      },
+      study_mode: {
+        open: { full_time: 1, part_time: 2, full_time_or_part_time: 3 },
+        closed: { full_time: 0, part_time: 0, full_time_or_part_time: 0 },
+      },
+      qualification: {
+        open: {
+          qts: 1, pgce_with_qts: 2, pgde_with_qts: 3, pgce: 4, pgde: 5
+        },
+        closed: {
+          qts: 0, pgce_with_qts: 0, pgde_with_qts: 0, pgce: 0, pgde: 0
+        },
+      },
+      is_send: {
+        open: { yes: 1, no: 2 },
+        closed:  { yes: 0, no: 0 },
+      },
+      subject: {
+        open: Subject.active.each_with_index.map { |sub, i| x = {}; x[sub.subject_name] = (i + 1) * 3; x }.reduce({}, :merge),
```

Comment:
i think you cheated and copied the implementation in to the test. why did you choose to do that here and have the actual outputs on the other parts of the test?

Reply:
The whole test is a cheat...

The input is a fake the in between is also a fake. 

The bit I'm really testing for in this whole thing is really


```ruby

{
 "I want this key to be transformed" : "either a number or defaulted as 0" 
} 

```